girder/_plugin.py

```diff
@@ -0,0 +1,266 @@
+        return
+
+    _pluginRegistry = {}
+    for entryPoint in iter_entry_points(NAMESPACE):
+        pluginClass = entryPoint.load()
+        plugin = pluginClass(entryPoint)
+        _pluginRegistry[plugin.name] = plugin
+
+
+def _walkPluginTree(func, plugins, dependencyGetter, handled, nodes):
+    """Walk through the plugin dependency tree depth first.
+
+    This is a recursive function performing a specialized topological sorting of the
+    plugin dependency tree.  Each time a new plugin is encountered the provided function
+    is called on the plugin config object.  This function will be called at most once
+    for each plugin and if it raises any exception, an error message will be recorded and
+    the walk cancelled.
+
+    :param func: A function taking the plugin configuration object.
+    :param plugins: A list of plugin names to traverse.
+    :param dependencyGetter: A function that returns a list of dependencies of a plugin.
+    :param handled: A set of plugin names that have already been handled.
+    :param nodes: A set of plugin names in the current tree traversal.  This is used
+                  to detect cycles in the dependency graph.
+    """
+    # loop over a sorted list of plugins to ensure a consistent traversal order
+    for pluginName in sorted(plugins):
+        if pluginName in handled:
+            continue
+
+        # handle dependency cycles
+        if pluginName in nodes:
+            logprint.error('Cyclic dependencies encountered while processing plugins')
+            raise Exception('Cyclic dependencies encountered while processing plugins')
+
+        # generate a set of nodes traversed up to the root for cycle detection
+        pluginNodes = {pluginName}
+        pluginNodes.update(nodes)
+
+        plugin = getPlugin(pluginName)
+
+        # handle missing dependencies
+        if plugin is None:
+            logprint.error('ERROR: Plugin %s not found', pluginName)
+            raise Exception('Required plugin missing')
+
+        # traverse into sub-dependencies
+        try:
+            _walkPluginTree(func, dependencyGetter(plugin), dependencyGetter, handled, pluginNodes)
+        except Exception:
+            logprint.error('ERROR: Dependency failure while processing %s', pluginName)
+            _pluginFailureInfo[pluginName] = {
+                'traceback': '',
+                'message': 'Dependency error'
+            }
+            handled.add(pluginName)
+            raise
+
+        # call the handler function and record it
+        func(plugin)
+        handled.add(pluginName)
+
+
+def _getToposortedPlugins(plugins=None, dependencyGetter=None):
+    """Return a toposorted list of plugins with a custom child-node accessor.
+
+    :param plugins: A list of plugins to include (defaults to all plugins)
+    :param dependencyGetter: A function taking a plugin object and returning a list of dependencies
+    """
+    dependencyGetter = dependencyGetter or (lambda p: p.dependencies)
```

Comment:
I wanted to avoid exposing the `dependencyGetter` argument publicly because it exposes too much of the internal implementation that we may want to replace.

For the `plugins` argument, do you mean duplicate the
```python
if plugins is None:
    plugins = allPlugins()
```
in the public functions?  I could do that, but I don't see a real benefit.

Reply:
I was not suggesting exposing dependencyGetter in getToposortedPlugins or getToposortedWebDependencies.  Only that you might move the default functionality defined in _getToposortedPlugins  into getToposortedPlugins:

```python
def getToposortedPlugins(plugins=None):
     """Return a toposorted list of plugins.
 
     By default this function will return a toposorted list of all detected plugins.  If a list of
     plugin names is provided, it will only return a minimal list of plugins required for the
     given top-level plugins.
     """
     return _getToposortedPlugins(
         plugins,
         dependencyGetter=lambda p: p.dependencies
    )
```

Equally you might consider moving the "defaulting" functionality of ```plugins``` into the public functions. That way _getToposortedPlugin  accepts plain arguments rather than keyword arguments and is only responsible for doing its work.  This makes testing the "defaulting" behavior explicitly apart of the public API and an easier target for testing.  But also... it is just a nitpick :smile: